### PR TITLE
feat(pipeline): enrich core pipelines processors.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Add three more types pipeline processors.

- `DuplicateProcessor`: duplicate the data from one input port to two output ports.
- `ShuffleProcessor`: reorder the data flow from input ports to output ports.
- Multiple producers single consumer sinker (`MpscSinker`, `AsyncMpscSinker`): sinker with multiple input ports.

Intuitive graphs:

```
Duplicate processor:

         ┌───────────┐
         │           ├─────► b
b ──────►│ Duplicate │
         │           ├─────► b.clone()
         └───────────┘


ShuffleProcessor:

       ┌─────────┐
b0────►│         ├───►b0
       │         │
b1────►│         ├───►b2
       │ Shuffle │
b2────►│[0,2,1,3]├───►b1
       │         │
b3────►│         ├───►b3
       └─────────┘

MpscSinker, AsyncMpscSink:

            ┌─────────────────┐
input0─────►│                 │
            │                 │
input1─────►│    MpscSinker   │
        .   │        /        │
        .   │ AsyncMpscSinker │
        .   │                 │
inputN─────►│                 │
            └─────────────────┘
```

### CODE REVIEW

File: src/query/pipeline/core/src/pipeline.rs [hunk 1]
```diff
@@ -210,6 +212,53 @@ impl Pipeline {
...
+                        vec![output1, output2],
+                    ));
+                }
+                self.pipes.push(Pipe::create(
+                    pipe.output_length,
+                    pipe.output_length * 2,
+                    items,
+                ));
+            }
+            _ => {}
```
Comment (zhang2014) [inline src/query/pipeline/core/src/pipeline.rs:238]: Need err

File: src/query/pipeline/core/src/pipeline.rs [hunk 2]
```diff
@@ -210,6 +212,53 @@ impl Pipeline {
...
+        }
+    }
+
+    pub fn shuffle(&mut self, rule: Vec<usize>) {
```
Comment (zhang2014) [inline src/query/pipeline/core/src/pipeline.rs:242]:
```suggestion
    pub fn reorder_inputs(&mut self, new_order: Vec<usize>) {
```

File: src/query/pipeline/core/src/processors/shuffle_processor.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+/// `rule` is a vector of [usize], each element is the index of the output port.
+///
+/// For example, if the rule is `[1, 2, 0]`, the data flow will be:
+///
+/// - input 0 -> output 1
+/// - input 1 -> output 2
+/// - input 2 -> output 0
+pub struct ShuffleProcessor {
+    inputs: Vec<Arc<InputPort>>,
+    outputs: Vec<Arc<OutputPort>>,
```
Comment (zhang2014) [inline src/query/pipeline/core/src/processors/shuffle_processor.rs:36]:
```suggestion
    channel: Vec<(Arc<InputPort>,Arc<OutputPort>)>,
```

File: src/query/pipeline/core/src/processors/shuffle_processor.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+impl Processor for ShuffleProcessor {
+    fn name(&self) -> String {
+        "Shuffle".to_string()
+    }
+
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn event(&mut self) -> Result<Event> {
```
Comment (zhang2014) [inline src/query/pipeline/core/src/processors/shuffle_processor.rs:126]: Fix the order when creating processor and finish processor when all channel are finished

File: src/query/pipeline/core/src/processors/duplicate_processor.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            return Ok(Event::Finished);
+        }
+        if self.input.is_finished() {
+            self.output1.finish();
+            self.output2.finish();
+            return Ok(Event::Finished);
+        }
+        if self.input.has_data() {
+            let block = self.input.pull_data().unwrap();
+            self.output1.push_data(block.clone());
```
Comment (zhang2014) [inline src/query/pipeline/core/src/processors/duplicate_processor.rs:68]: output may cannot push

File: src/query/pipeline/sinks/src/async_mpsc_sink.rs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    #[unboxed_simple]
+    async fn consume(&mut self, data_block: Vec<DataBlock>) -> Result<bool>;
+}
+
+pub struct AsyncMpscSinker<T: AsyncMpscSink + 'static> {
+    inner: T,
+    finished: bool,
+    inputs: Vec<Arc<InputPort>>,
+    inputs_data: Option<Vec<DataBlock>>,
```
Comment (zhang2014) [inline src/query/pipeline/sinks/src/async_mpsc_sink.rs:47]:
```suggestion
    inputs_data: Option<DataBlock>,
```

Comment (zhang2014) [inline src/query/pipeline/sinks/src/async_mpsc_sink.rs:47]:
```suggestion
    inputs_data: Option<DataBlock>,
```

Comment (RinChanNOWWW) [reply src/query/pipeline/sinks/src/async_mpsc_sink.rs:47]: Why not consume a batch of `DataBlock`s one time?

Comment (zhang2014) [reply src/query/pipeline/sinks/src/async_mpsc_sink.rs:47]: We only need to consume all inputs once in first event.

File: src/query/pipeline/core/src/processors/duplicate_processor.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            return Ok(Event::Finished);
+        }
+
+        if self.input.is_finished() {
+            self.output1.finish();
+            self.output2.finish();
+            return Ok(Event::Finished);
+        }
+
+        if !can_push1 || !can_push2 {
```
Comment (zhang2014) [inline src/query/pipeline/core/src/processors/duplicate_processor.rs:83]: no any buffer?

File: src/query/pipeline/core/src/processors/shuffle_processor.rs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        let mut finished = Vec::with_capacity(self.not_finished.len());
+        let mut need_data = true;
+        for i in self.not_finished.iter() {
```
Comment (zhang2014) [inline src/query/pipeline/core/src/processors/shuffle_processor.rs:54]:
```suggestion
        for (input, output) in &self.channel {
```

**Review by ariesdevil — APPROVED** [review]

**Pull request merged**